### ISSUE TITLE
Deadline Nuke CopyCat Distributed Training Fails with c10d error

### ISSUE THREAD

**Issue by jin-raii** [banjac90/CopyCat-Nuke-Submitter#1]
Hello there!!

First off, Great Plugin!!

However, I'm having difficulties in Distributed Environment. the job starts but doesn't decrease the time it was supposed to or something I missed don't know... help me out here?

some info about the setup....

**Environment:**

Deadline Version: 10
Nuke Version: 15.1v4 

Main Machine & Worker Logs ( portion ) : 

> Checking memory .1.2.3.4.5.6.7.8.9
> Process of rank 1 is connecting to (192.168.214.20,3000).
> STDOUT: [W socket.cpp:663] [c10d] The client socket has failed to connect to [HOSTNAME.xxx.local]:3000 (system error: 10049 - unknown error).
> Process of rank 1 is waiting for other processes to join.
> STDOUT: Global rank: 1, world size: 2
>  STDOUT: Validating Inputs .1.2.3.4.5.6.7.8.9
>  Caching Inputs .1.2.3.4.5.6.7.8.9
> 2025-04-24 17:22:28:  0: STDOUT: Checking memory .1.2.3.4.5.6.7.8.9
> 2025-04-24 17:22:28:  0: STDOUT: Checking memory .1.2.3.4.5.6.7.8.9
> 2025-04-24 17:22:29:  0: STDOUT: Checking memory .1.2.3.4.5.6.7.8.9
> 2025-04-24 17:22:34:  0: STDOUT: Estimating training time .3.6.9[Step:1/163636] Loss:0.00885404
> 2025-04-24 17:22:34:  0: STDOUT: [Step:10/163636] Loss:0.000274674
> 2025-04-24 17:22:34:  0: STDOUT: [Step:20/163636] Loss:0.000116908
> 2025-04-24 17:22:35:  0: STDOUT: [Step:30/163636] Loss:0.000133878
> 2025-04-24 17:22:35:  0: STDOUT: [Step:40/163636] Loss:6.11843e-05
> 2025-04-24 17:22:36:  0: STDOUT: [Step:50/163636] Loss:2.76682e-05

In the distributed training setup, it seems the hostname (Machine.xxx.local) resolves to an IPv6 address, which might be causing mismatches and preventing gradient sharing at the SyncInterval. Is there a way to configure the plugin to explicitly use IPv4 or avoid hostname resolution conflicts?

BTW, I followed all the docs to setup CopyCat plugin in Deadline Render Farm.

Any help would be appreciated.
Thank you

**Comment by banjac90** [comment]
Hello,

Apologies for the delayed response — I hope it's not too late!

I've just pushed a fix to ensure the plugin now consistently prioritizes and retrieves the IPv4 address by default.

Additionally, the following validation logic is in place:
```
 if thisMachine == mainmachine and ipAddress != mainMachineIp:
            self.FailRender("Your Main Machine IP is incorrect! Please check main machine IP!")

```
This means the plugin checks whether the current machine's name and IP match the expected main machine configuration. If there’s a mismatch, it will raise a FailRender error.

In your case, these checks are passing, so everything should be working correctly.

Please try the updated plugin and let me know if the issue persists.

Best regards!

**Comment by jin-raii** [comment]
Thanks for the reply,

We found that the error I mentioned has nothing to do with the code (your code is solid). `The get_local_ip()` function does return an IPv4 address.

The training does start, but it takes twice as long compared to running on a single machine.

Do you have any working examples or resources where distributed training actually helps decrease training time?

Anyway, thanks again for the commit!

**Comment by craterstudio** [comment]
Hello,

Thanks for the update!

To be honest, we don’t have a definitive benchmark yet showing consistent time savings from distributed training. We’ve done some training within Nuke, and while it doesn’t always speed up the process, it does free up our artists’ workstations—which is a big win on its own.

We’ve observed that our RTX 4090 GPUs perform better when training together in a dedicated pool—without being mixed with other machines or different GPU models (like the RTX 30xx series). When using only 4090s, the training process does speed up noticeably. So, from my limited experience, I’d say hardware compatibility is likely what affects performance the most.

Out of curiosity—are your render nodes equipped with high-end GPUs? That could be a key factor in the behavior you're seeing.

**Comment by banjac90** [comment]
Hello,

Thanks for the update!

To be honest, we don’t have a definitive benchmark yet showing consistent time savings from distributed training. We’ve done some training within Nuke, and while it doesn’t always speed up the process, it does free up our artists’ workstations—which is a big win on its own.

We’ve observed that our RTX 4090 GPUs perform better when training together in a dedicated pool—without being mixed with other machines or different GPU models (like the RTX 30xx series). When using only 4090s, the training process does speed up noticeably. So, from my limited experience, I’d say hardware compatibility is likely what affects performance the most.

Out of curiosity—are your render nodes equipped with high-end GPUs? That could be a key factor in the behavior you're seeing.

**Comment by jin-raii** [comment]
Hello,

Our pool consists of RTX 4070 Ti and RTX 5090 GPUs.

We tried using three machines, each with the same hardware configuration (RTX 4070 Ti). As you mentioned, this setup does free up our artists' workstations, but the training time takes twice as long compared to using a single machine, as I mentioned above.

Below is some additional information:

```
Dataset size: 30
Epochs: 60,000

Single Machine:

Batch Size: 10

Sync Interval: 180,000

Training Time: 3 hours 7 minutes

3 Machines:

Batch Size: 60

Sync Interval: 1,000

Steps: 30,000

Training Time: 7 hours 5 minutes
```

Have you encountered any issues like this?

**Comment by banjac90** [comment]
Hello,

Thank you for the update.

I tried running a similar test on my side using a smaller model and dataset to compare results.
```
Dataset size: 25
Epochs: 60,000

Single Machine:

Batch Size: 10
Sync Interval: 180,000
Training Time: 2 hours 57 minutes

```
Distributed (3 Machines):
```
Batch Size: 60
Sync Interval: 1,000
Steps: 60,000
Training Time: 1 hour 25 minutes

```

Based on these results, distributed training did speed things up in my case.

But I also encountered a failed job, where one machine took significantly longer to complete training. After investigating, I discovered it was using the CPU instead of the GPU. :)

Here are a few things I recommend checking:

- Ensure that all three machines are actively using their GPUs during training. 
- Make sure the NVIDIA drivers are up to date on all machines.
- Check and clear the Nuke cache on each machine, if possible.

Most likely in your case is CPU Fallback. 

Are you able to monitor your machines during training to confirm that they’re using the GPU for rendering?

**Comment by jin-raii** [comment]
Hello,

Thank you for testing — I really appreciate it! I also ran a training session, this time with `CUDA_CACHE_MAXSIZE=4294967296`, and I believe the training time dropped significantly — from around 7 hours to 56 minutes and 50 seconds. I used the same Nuke script, with a dataset size of 30.

This was my job config: 

```
BatchMode=False
BatchModeIsMovie=False
ContinueOnError=False
CopyCatNode=CopyCat3
EnforceRenderOrder=False
GpuOverride=0
MainMachine=pc1
MainMachineIP=192.168.xx.xx
Port=3000
SceneFile=script_test.nk
SyncInterval=1000
TrainingSlaves=pc1,pc2,pc3
UseGpu=True
UseIPv6=False
UseSpecificGpu=False
Version=15.1
WorldSize=3
```

Just wanted to check — how do you decide on the batch size relative to the dataset size? Is there a preferred method or guideline you follow for setting that?

Thanks again!

**Comment by banjac90** [comment]
Hello,

Honestly, the batch size is usually chosen somewhat randomly by the artists :) — it’s mostly a matter of trial and error.

My main goal was to train a model that closely resembles the results you achieved, so I focused more on matching your batch size and dataset size, also this has overfitting result at the end, on our data :)
